### Задание 1.
- Скачать файл в уроке 
- Считать данные с помощью pandas
- Вывести на экран первые 5 строк
- Посмотреть на описание признаков и на их содержание

In [45]:
import pandas as pd

In [46]:
import pathlib
from pathlib import Path

In [47]:
work_path = pathlib.Path.cwd()
data_path = Path(work_path, 'data', 'kc_house_data.csv')

In [48]:
df = pd.read_csv(data_path)

In [49]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living   
0  7129300520  20141013T000000  221900.0         3       1.00         1180  \
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement   
0      5650     1.0           0     0  ...      7        1180              0  \
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15   
0      1955             0    98178  47.5112 -122.257           1340  \
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

### Задание 2.
Проведите первичный анализ данных:
- Изучите типы данных
- Найдите количество пропущенных ячеек в данных
- Посчитайте основные статистики по всем признакам и поизучайте их
- Пишите выводы

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [51]:
df.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [52]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living   
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000  \
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition   
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000  \
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated   
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000  \
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

#### Выводы:
- В базе данных 21613 строки;
- В базе данных нет пропущенных ячеек;
- Это база данных объектов недвижимости;
- Данные по объектам недвижимости собраны по 21 параметру, пропусков нет, данные в основном числовые, только в столбце дата строка;
- 

### Задание 3.
Ответьте на несколько вопросов
- 3.1 В каком диапазоне изменяются стоимости недвижимости?
- 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?
- 3.3 Как много домов с разными этажами в данных?
- 3.4 Насколько хорошие состояния у домов в данных?
- 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?


In [53]:
min(df['price']), max(df['price'])

(75000.0, 7700000.0)

In [74]:
living = df['sqft_living'].mean()
print(living)
lot = df['sqft_lot'].mean()
print(lot)
living_share = living / (lot / 100)
print(living_share)

2079.8997362698374
15106.967565816869
13.767817579592272


In [55]:
lot = df['sqft_lot'].mean()
print(lot)

15106.967565816869


In [56]:
df['floors'].value_counts()

floors
1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: count, dtype: int64

In [57]:
df['condition'].value_counts()

condition
3    14031
4     5679
5     1701
2      172
1       30
Name: count, dtype: int64

In [58]:
min(df['yr_built']), max(df['yr_built'])

(1900, 2015)

### Задание 4.
Ответьте на несколько вопросов

- 4.1 Сколько в среднем стоят дома, у которых 2 спальни?
- 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?
- 4.3 Как много домов коснулся ремонт?
- 4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [65]:
df[df['bedrooms'] == 2]['price'].mean()

401372.681884058

In [64]:
df[df['price'] > 600000]['sqft_lot'].mean()

20442.524776214832

In [73]:
df_new = df[df['yr_renovated'] > 0]
df_new.shape[0]

914

In [66]:
df[df['grade'] > 10]['price'].mean()

1678635.1175298805

In [67]:
df[df['grade'] < 4]['price'].mean()


189750.0

### Задание 5.
Ответьте на несколько вопросов
- 5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?
- 5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?
- 5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [87]:
df_new = df[(df['waterfront'] == 1) & (df['bedrooms'] >= 3) & (df['sqft_basement'] > 0)]
df_new

id             date      price  bedrooms  bathrooms   
49      822039084  20150311T000000  1350000.0         3       2.50  \
246    2025069065  20140929T000000  2400000.0         4       2.50   
300    3225069065  20140624T000000  3075000.0         4       5.00   
457    2122039094  20141126T000000   705000.0         3       3.00   
656    3760500116  20141120T000000  3070000.0         3       2.50   
...           ...              ...        ...       ...        ...   
18914  9253900408  20150408T000000  1400000.0         3       2.75   
20325   518500480  20140811T000000  3000000.0         3       3.50   
20767  8043700105  20150417T000000  2300000.0         4       4.00   
21201   518500460  20141008T000000  2230000.0         3       3.50   
21576  9253900271  20150107T000000  3567000.0         5       4.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade   
49            2753     65005     1.0           1     2  ...      9  \
246           3650      8354     1.0           1     4  ...      9   
300           4550     18641     1.0           1     4  ...     10   
457           1970     20978     2.0           1     3  ...      9   
656           3930     55867     1.0           1     4  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
18914         3130     19530     1.0           1     4  ...      8   
20325         4410     10756     2.0           1     4  ...     11   
20767         4360      8175     2.5           1     4  ...     10   
21201         3760      5634     2.0           1     4  ...     11   
21576         4850     10584     2.0           1     4  ...     10   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat   
49           2165            588      1953             0    98070  47.4041  \
246          1830           1820      2000             0    98074  47.6338   
300          2600           1950      2002             0    98074  47.6053   
457          1770            200      1980             0    98070  47.3844   
656          2330           1600      1957             0    98034  47.7022   
...           ...            ...       ...           ...      ...      ...   
18914        1690           1440      1947          1984    98008  47.5895   
20325        3430            980      2014             0    98056  47.5283   
20767        3940            420      2007             0    98008  47.5724   
21201        2830            930      2014             0    98056  47.5285   
21576        3540           1310      2007             0    98008  47.5943   

          long  sqft_living15  sqft_lot15  
49    -122.451           2680       72513  
246   -122.072           3120       18841  
300   -122.077           4550       19508  
457   -122.438           2280       75396  
656   -122.224           2730       26324  
...        ...            ...         ...  
18914 -122.111           2980       18782  
20325 -122.205           3550        5634  
20767 -122.104           2670        8525  
21201 -122.205           3560        5762  
21576 -122.110           3470       18270  

[84 rows x 21 columns]

In [106]:
df_new = df[(df['waterfront'] == 1) | (df['view'] == df['view'].max())]
df_new_1 = df_new[df_new['yr_built'] >= 1980]
df_new_1['price'].min(), df_new_1['price'].max()

(380000.0, 6885000.0)

In [108]:
df_new = df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] <= 150000)]
df_new['condition'].mean()

3.0